### ISSUE TITLE
Langchain Expression Language (LCEL) pass through does not work with two consecutive chain

### ISSUE THREAD

**Issue by SoulEvill** [langchain-ai/langchain#18173]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.

### Example Code

There are cases where user needs to pass through variables for more than one chains for later use, but current implementation doesnt support this. 

Reproducible example following on RAG langchain expression language example from https://python.langchain.com/docs/expression_language/cookbook/retrieval

`from operator import itemgetter

from langchain_community.vectorstores import FAISS
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import RunnableLambda, RunnablePassthrough
from langchain_openai import ChatOpenAI, OpenAIEmbeddings

vectorstore = FAISS.from_texts(
    ["harrison worked at kensho"], embedding=OpenAIEmbeddings()
)
retriever = vectorstore.as_retriever()

template = """Answer the question based only on the following context:
{context}

Question: {question}
"""
prompt = ChatPromptTemplate.from_template(template)

model = ChatOpenAI()

chain = (
    {"context": retriever, "question": RunnablePassthrough()}
    ### Only line added to the example
    | {'context': itemgetter('context'), "question": itemgetter('question')}
    | prompt
    | model
    | StrOutputParser()
)

chain.invoke("where did harrison work?")
'

### Error Message and Stack Trace (if applicable)

Traceback (most recent call last):
  File "/Users/zhengisamazing/1.python_dir/vigyan-llm-api/dev/langchain_playground.py", line 110, in <module>
    chain.invoke("where did harrison work?")
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 2056, in invoke
    input = step.invoke(
            ^^^^^^^^^^^^
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 2693, in invoke
    output = {key: future.result() for key, future in zip(steps, futures)}
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 2693, in <dictcomp>
    output = {key: future.result() for key, future in zip(steps, futures)}
                   ^^^^^^^^^^^^^^^
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 3504, in invoke
    return self._call_with_config(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 1243, in _call_with_config
    context.run(
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/site-packages/langchain_core/runnables/config.py", line 326, in call_func_with_variable_args
    return func(input, **kwargs)  # type: ignore[call-arg]
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 3378, in _invoke
    output = call_func_with_variable_args(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhengisamazing/opt/anaconda3/envs/vigyan-llm-api/lib/python3.11/site-packages/langchain_core/runnables/config.py", line 326, in call_func_with_variable_args
    return func(input, **kwargs)  # type: ignore[call-arg]
           ^^^^^^^^^^^^^^^^^^^^^
TypeError: string indices must be integers, not 'str'

### Description

There are cases where user needs to pass through variables for more than one chains for later use, but current implementation doesnt support this. 

Provided reproducible example following on RAG langchain expression language example from https://python.langchain.com/docs/expression_language/cookbook/retrieval

### System Info

langchain==0.1.7
langchain-cli==0.0.21
langchain-community==0.0.20
langchain-core==0.1.27
langchain-google-genai==0.0.9
langchain-openai==0.0.6

platform: mac
python version:3.11.7

**Comment by hinthornw** [comment]
The reason this doesn't work is because python parses the AST in left to right order. You are piping two stdlib python dicts together before it touches any langchain code, which implicitly deletes the first dict. 

See for yourself:

```
{"context": retriever, "question": RunnablePassthrough()} | {
    "context": itemgetter("context"),
    "question": itemgetter("question"),
}
```
Results in output:
```
{'context': operator.itemgetter('context'),
 'question': operator.itemgetter('question')}
```

To fix, explicitly create the langchain object using  RunnableParallel:
```
RunnableParallel({"context": retriever, "question": RunnablePassthrough()})
### Only line added to the example
| {'context': itemgetter('context'), "question": itemgetter('question')}
| prompt
| model
| StrOutputParser()
)
```

Then the first function in the sequence is a langchain object, which can be composed with dicts, runnables, etc. as intended.

**Comment by yfontana** [comment]
I ran into a similar issue, and it took me a while to figure out that I needed to replace dicts with RunnableParallel.
I would suggest either:
- Updating the docs and examples to make this behavior clear and explicit (particularly the "TIP" section on https://python.langchain.com/docs/expression_language/primitives/parallel/, which states that a dict and a RunnableParallel are equivalent. In this case they aren't.)
- Finding a way to make this syntax work. It is easy to run into this as soon as you want to implement a somewhat complex chain, and it's not intuitive for a non-Python expert that a dict would work for one step but not for a second one.
